### ISSUE TITLE
Install SharePoint 2016 with Feature Pack 1 Min Roles: Application with Search, WFE with Distributed Cache

### ISSUE THREAD

**Issue by andikrueger** [PowerShell/SharePointDsc#579]
**Details of the scenario you try and problem that is occurring:**
With SharePoint Feature Pack two new Min Roles are available. Currently it is not possible to install a new SharePoint Farm with these two roles as a current CU is required to enable the new Role Switches. 
As far as I see it's not possible to install an update prior to farm creation. The Config DB Parameter is mandatory for the SPProductUpdate.

**The DSC configuration that is using the resource:**
N/A
**Version of the Operating System and PowerShell the DSC Target Node is running:**
N/A
**Version of the DSC module you're using:**
1.6

**Comment by BrianFarnhill** [comment]
Hi @andikrueger 

It very much so is possible to install a new farm with those roles. You install the update by either slipstreaming it in to your binaries directory, or by installing the binary part of the update after you install SharePoint but before you create/join a server to the farm.

The part that isn't possible is to use the SPProductUpdate resource - due to the way that we can tell if an update is installed or not it relies on the config DB to do this. So if you want to install as part of a standard install, I recommend you slip stream the update as this will install it automatically when the main binaries are installed. 

There is a good example of how to Slipstream updates on Todd Klindt's blog at http://www.toddklindt.com/blog/Lists/Posts/Post.aspx?ID=403 which should get you started.

I hope this helps, let me know if you have any more questions. I'll close this one off as the issue you are seeing there is by design but the slipstreaming is a work around.

**Comment by andikrueger** [comment]
Hi @BrianFarnhill,
thank you for the quick response. 
First, I thought about slipstreaming the updates, but then postponed it due to the need to install language packs for SharePoint too. AFAIK language packs should be installed prior to the CU, as the CU holds language dependent files which would not be updated otherwise. [e.g. For SharePoint 2013](https://sharepoint.stackexchange.com/questions/136770/what-is-the-installation-order-to-add-new-language-pack-to-an-existing-sharepoin)

Doing it the DSC way without a slipstreamed version of SharePoint would need a bit more of configuration. I would use the Package and the xHotfix Resource after SPInstall:

For the Language Pack(s):
``
Package InstallSharePointLanguagePack
{
	 Ensure = 'Present';	PsDscRunAsCredential = $SPAdmin; Name = ''; Path = ''; Arguments = '/quiet'; ProductId = ''
}``
Name, Path and ProductId could be obtained from the language pack installer with a tool like ORCA (Part of the Windows SDK for Windows Installer Developers)...

For the CU:
The CU part looks a bit trickier to me, as this is a real update. Maybe xHotfix from [xWindowsUpdate](https://github.com/PowerShell/xWindowsUpdate) could to it. 

Do you think this approach is valid?

**Comment by ykuijs** [comment]
Hi @andikrueger, you are absolutely correct. The sliptstream method doesn't work that well if you have to use language packs. However, you can combine both methods: Slipstream to fix your issue, then install the language packs and then use SPProductUpdate to install the CU again, which will then only update the language packs (since SP is already on that patch level).

Or slipstream the Nov 16 update (FP1) and then use SPProductUpdate to install the most recent CU to update all components to the most recent patch level.

**Comment by andikrueger** [comment]
Hi @ykuijs, 
I really like your idea. It is a good to slipstream an older CU (like the Nov 16 one) and continue from that point with the language packs and a newer CU. In this case, the newer CU should install fine and the language pack becomes updated too.
I always thought about a slipstream version with the latest CU. Recently I experienced some strange issues in that case when I tried to apply the latest CU again. Most of the time the CU installer started and didn't do anything, as the CU was no longer suitable for my environment.

Using an older CU will result in the following prerequisite and sequence for the installation:
prerequisite: Slipstream SharePoint 2016 + Nov 16 CU
- Install Farm
- Create Farm (New Min Roles)
- Install Language Packs
- Upgrade to the latest CU

This is a good chance to get a proper Min Role 2016 SharePoint Farm with Language Packs.

**Comment by andikrueger** [comment]
Hi @ykuijs ,
just a followup on this topic: Yesterday I ran into an issue with SPProductUpdate. Starting with a slipstreamed SharePoint 2016 with the Sept 2017 CU and with the following steps:

- Install Farm
- Create Farm
- Install Language Pack
- Run SPPorduct Update

will result in an error:
```
SharePoint update install failed, exit code was 17025. Error codes can be found at https://aka.ms/installerrorcodes"
```
Code 17025 is "Patch already installed". Which is fine in that case, but horrible for the whole configuration, cause it will not complete.

Adding the Exit Code 17025 to the [switch statement]( https://github.com/PowerShell/SharePointDsc/blob/5d9c45939afe422846c4050dd1e5a8572017b42b/Modules/SharePointDsc/DSCResources/MSFT_SPProductUpdate/MSFT_SPProductUpdate.psm1#L429) in the MSFT_SPPRoductUpdate would solve this issue.

**Comment by ykuijs** [comment]
Hi @andikrueger, I do not understand why it would try to install the CU at all. The Test-TargetResource would have to detect that SharePoint is already on the correct build number and never try to install the update.
